### PR TITLE
view generated rainlang

### LINKED ISSUE #458: ability to see the result of parsing while editing dotrain in app
currently there's now way to see/preview/review the actual parsed code from the dotrain scenario that will be deployed and/or is being used to produce charts

at the least this makes it unnecessarily difficult to figure out the right stack trace to use, which is error prone as it requires mentally reproducing the topological sort

### PR DESCRIPTION
**Pull request by rouzwelt**
resolves #458

### CODE REVIEW

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 1]
```diff
@@ -161,8 +212,46 @@
...
-<Charts {chartData} />
+<Tabs
+  style="underline"
+  contentClass="mt-4"
+  defaultClass="flex flex-wrap space-x-2 rtl:space-x-reverse mt-4"
+>
+  <TabItem open title="Rainlang">
+    <div class="flex flex-col gap-y-2">
+      <Label>Select Scenario</Label>
+      {#if scenarios === undefined || Object.keys(scenarios).length === 0}
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:223]: i'm getting concerned about all the different ways we check for the lack of a value across this codebase, coupled with the lack of tests, and observance of real bugs in the wild due to this fact

can we defer to a lib like lodash to handle these subtle yet repetitive checks?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: tauri-app/src/routes/orders/add/+page.svelte (hunk 1 above)
Comment (rouzwelt) [reply tauri-app/src/routes/orders/add/+page.svelte:223]: Im not sure I understand exactly what you're asking, `scenarios`here  uses lodash `pickedBy()`which can result in undefined or empty object if there are no scenarios defined, and here then it is just checked if it's empty or undefined.

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 2]
```diff
@@ -161,8 +212,46 @@
...
+  <TabItem open title="Rainlang">
+    <div class="flex flex-col gap-y-2">
+      <Label>Select Scenario</Label>
+      {#if isEmpty(scenarios)}
+        <span class="text-gray-500 dark:text-gray-400">No scenarios found for the selected network</span>
+      {:else}
+        <div class="flex justify-end gap-x-2">
+          <DropdownRadio options={scenarios} bind:value={scenarioRef}>
+            <svelte:fragment slot="content"  let:selectedRef>
+              <span>{selectedRef !== undefined ? selectedRef : 'Select a scenario'}</span>
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:229]: lodash check here

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 3]
```diff
@@ -103,27 +139,43 @@
...
+
+  async function generateRainlangString() {
+    try {
+      if(!scenario) throw Error("Select a scenario to generate rainlang");
+      rainlangText = await orderAddComposeRainlang($dotrainFile.text, scenario);
+      resetRainlang = false;
+      // eslint-disable-next-line no-empty
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log(e);
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:151]: this needs to send errors to sentry with #494

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 4]
```diff
@@ -3,62 +3,98 @@
...
-    completion: (text, position) => completionCallback.apply(null, [text, position, bindings]),
   });
 
   $: {
     if(deploymentRef === undefined && deployments !== undefined && Object.keys(deployments).length > 0) {
       deploymentRef = Object.keys(deployments)[0];
     }
   }
+  $: {
+    if(scenarioRef === undefined && scenarios !== undefined && Object.keys(scenarios).length > 0) {
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:92]: lodash check

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 5]
```diff
@@ -3,62 +3,98 @@
...
+        ];
+      }
+      try {
+        // get problems with dotrain
+        problems = await promiseTimeout(
+          problemsCallback(text, bindings, deployment?.scenario.deployer.address),
+          5000,
+          "failed to parse on native parser"
+        );
+      } catch(e) {
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:75]: sentry

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 6]
```diff
@@ -3,62 +3,98 @@
...
+  $: scenario = (scenarioRef !== undefined && mergedConfig !== undefined) ? mergedConfig.scenarios[scenarioRef] : undefined;
 
+  const rainlangExtension = new RawRainlangExtension({
+    diagnostics: async (text) => {
+      let configProblems = [];
+      let problems = []
+      try {
+        // get problems with merging settings config with frontmatter
+        configProblems = await mergeDotrainConfigWithSettingsProblems(text.text);
+      } catch(e) {
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:59]: sentry

File: tauri-app/src/routes/orders/add/+page.svelte [hunk 7]
```diff
@@ -3,62 +3,98 @@
...
 
-  $: rainlangExtension = new RawRainlangExtension({
-    hover: (text, position) => hoverCallback.apply(null, [text, position, bindings]),
-    diagnostics: async (text) => {
-      // get problems with merging settings config with frontmatter
-      const configProblems = await mergeDotrainConfigWithSettingsProblems(text.text);
-
-      // get problems with dotrain
-      const problems = await problemsCallback.apply(null, [text, bindings, deployment?.scenario.deployer.address]);
+  $: scenarios = (mergedConfigSource !== undefined && mergedConfigSource?.scenarios !== undefined) ?
```
Comment (thedavidmeister) [inline tauri-app/src/routes/orders/add/+page.svelte:48]: lodash

File: tauri-app/src/lib/components/ModalExecute.svelte [hunk 8]
```diff
@@ -44,7 +44,7 @@
       bind:derivationIndex={$ledgerWalletDerivationIndex}
       bind:walletAddress={$ledgerWalletAddress.value}
     />
-    <ButtonLoading on:click={() => executeLedger().finally(() => reset())} disabled={isSubmitting || !$ledgerWalletAddress || $ledgerWalletDerivationIndex == undefined || $ledgerWalletDerivationIndex == null} loading={isSubmitting}>
+    <ButtonLoading class="w-full" on:click={() => executeLedger().finally(() => reset())} disabled={isSubmitting || !$ledgerWalletAddress || $ledgerWalletDerivationIndex == undefined || $ledgerWalletDerivationIndex == null} loading={isSubmitting}>
```
Comment (thedavidmeister) [inline tauri-app/src/lib/components/ModalExecute.svelte:47]: lodash

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: tauri-app/src/routes/orders/add/+page.svelte (hunk 5 above)
Comment (rouzwelt) [reply tauri-app/src/routes/orders/add/+page.svelte:76]: this is not a sentry log, this is just a problem on lsp text area

**Review by thedavidmeister — APPROVED** [review]

**Pull request merged**